namedisl/__init__.py

```diff
@@ -32,72 +35,612 @@
 THE SOFTWARE.
 """
 
+import operator
 import re
-from collections.abc import Mapping
+from collections.abc import Callable, Mapping, Sequence
 from dataclasses import dataclass
 from importlib import metadata
-from typing import TypeAlias, TypeVar, overload
+from typing import Any, Generic, Self, TypeAlias, TypeVar, overload
 
 from constantdict import constantdict
 
 import islpy as isl
+from islpy import dim_type
 
 
 __version__ = metadata.version("namedisl")
 _match = re.match(r"^([0-9.]+)([a-z0-9]*?)$", __version__)
 assert _match
 VERSION = tuple(int(nr) for nr in _match.group(1).split("."))
 
+ALL_DIM_TYPES = [dim_type.param, dim_type.set, dim_type.in_, dim_type.div]
 
-IslObject = TypeVar("IslObject", isl.BasicSet, isl.Set)
+
+IslSetLikeObject = isl.Set | isl.Map
+
+IslExpressionLikeObject = isl.Aff | isl.QPolynomial
+IslPwExpressionLikeObject = isl.PwAff | isl.PwQPolynomial
+IslMultiExpressionLikeObject = isl.MultiAff | isl.PwMultiAff
+
+IslAllExpressionLike = (
+    IslExpressionLikeObject |
+    IslPwExpressionLikeObject |
+    IslMultiExpressionLikeObject
+)
+
+IslObject = IslSetLikeObject | IslAllExpressionLike
 NameToDim: TypeAlias = Mapping[str, tuple[isl.dim_type, int]]
 
+IslTypeT = TypeVar("IslTypeT", bound=IslObject)
+
+
+@dataclass(frozen=True)
+class NamedIslObject(Generic[IslTypeT]):
+    _obj: IslTypeT
+    _name_to_dim: NameToDim
+    # TODO: add cache
+
+    # FIXME: better name?
+    def dim_type_names(
+        self,
+        dim_type: dim_type
+    ) -> Sequence[str]:
+        return [
+            name
+            for name, (dt, _) in self._name_to_dim.items()
+            if dim_type == dt
+        ]
+
+    def dim(self, name: str) -> int:
```

Comment:
As above: The user really should not have to think about/have access to dim indices.